dali/pipeline/operator/sequence_shape.h

```diff
@@ -0,0 +1,298 @@
+
+#ifndef DALI_PIPELINE_OPERATOR_SEQUENCE_SHAPE_H_
+#define DALI_PIPELINE_OPERATOR_SEQUENCE_SHAPE_H_
+
+#include <memory>
+#include <string>
+#include <utility>
+#include "dali/pipeline/data/tensor.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+
+/**
+ * @brief Describes expandable prefix of the layout and of the shape of a batch.
+ */
+class ExpandDesc {
+ public:
+  inline ExpandDesc(const TensorListShape<> &shape, TensorLayout layout,
+                    bool should_expand_channels)
+      : layout_{layout},
+        frames_dim_{VideoLayoutInfo::FrameDimIndex(layout)},
+        channels_dim_{VideoLayoutInfo::ChannelDimIndex(layout)},
+        expand_frames_{frames_dim_ == 0 ||
+                       (should_expand_channels && frames_dim_ == 1 && channels_dim_ == 0)},
+        expand_channels_{should_expand_channels && 0 <= channels_dim_ && channels_dim_ <= 1},
+        num_expand_dims_{expand_channels_ + expand_frames_},
+        is_channel_first_{num_expand_dims_ == 2 && channels_dim_ < frames_dim_},
+        dims_to_expand_{shape.first(num_expand_dims_)},
+        num_expanded_{[&]() {
+          auto num_elements = dims_to_expand_.num_elements();
+          return static_cast<std::make_unsigned_t<decltype(num_elements)>>(num_elements);
+        }()} {}
+
+
+  inline bool ExpandChannels() const {
+    return expand_channels_;
+  }
+
+  inline bool ExpandFrames() const {
+    return expand_frames_;
+  }
+
+  inline bool IsChannelFirst() const {
+    return is_channel_first_;
+  }
+
+  inline int NumDimsToExpand() const {
+    return num_expand_dims_;
+  }
+
+  inline size_t NumExpanded() const {
+    return num_expanded_;
+  }
+
+  inline size_t NumSamples() const {
+    return dims_to_expand_.num_samples();
+  }
+
+  inline size_t NumExpanded(size_t sample_idx) const {
+    assert(sample_idx < NumSamples());
+    return volume(dims_to_expand_[sample_idx]);
+  }
+
+  inline size_t NumFrames(size_t sample_idx) const {
+    assert(sample_idx < NumSamples());
+    assert(frames_dim_ < NumDimsToExpand());
+    return dims_to_expand_[sample_idx][frames_dim_];
+  }
+
+  inline size_t NumChannels(size_t sample_idx) const {
+    assert(sample_idx < NumSamples());
+    assert(channels_dim_ < NumDimsToExpand());
+    return dims_to_expand_[sample_idx][channels_dim_];
+  }
+
+  inline const TensorListShape<> &DimsToExpand() const {
+    return dims_to_expand_;
+  }
+
+  inline TensorLayout Layout() const {
+    return layout_;
+  }
+
+  inline TensorLayout ExpandedLayout() const {
+    return layout_.first(num_expand_dims_);
+  }
+
+ private:
+  TensorLayout layout_;
+  int frames_dim_;
+  int channels_dim_;
+  bool expand_frames_;
+  bool expand_channels_;
+  int num_expand_dims_;
+  bool is_channel_first_;
+  TensorListShape<> dims_to_expand_;
+  size_t num_expanded_;
+};
+
+namespace sequence_utils {
+
+template <typename FrameRange>
+class SliceIterator {
+  using IndexType = typename FrameRange::IndexType;
```

Comment:
It could be FrameIterator or just Iterator if you feel some of those naming is more readable. it is used to create iterators that can unfold some number of outermost extents of the sample and iterate over the inner ones (so in case of FHWC sample and extending 1 extent you have an iterator that goes over HWC frames). It is a utility (used together with TensorVectorBuilder) solely for simplifying the code and reducing boilerplate.